fix(ext/node): improve shelljs compat with managed npm execution

kt3k opened the pull request:
This PR improves the Node.js compatibility in managed npm resolution mode.

Currently when an npm script executes the script next to it (`child.js`), the package.json next to it is discovered as config file for it. As a result, Deno uses `$DENO_DIR/npm/registry.npmjs.org/<package>/<version>/node_modules` as root of npm modules. As a result, the script (`child.js`) is considered Deno script, not a npm script. That leads to the error `ReferenceError: require is not defined`.

This PR fixes it by disabling the discovery of `node_modules` when the main specifier is inside of `DENO_DIR`.

closes #22732
closes #24589

dsherret commented on cli/factory.rs:
```diff
@@ -367,6 +367,19 @@ impl CliFactory {
+              }
+            } else {
+              Some(node_modules.clone())
+            });
```
As discussed in person: instead of doing this here, let's move this up into the `resolve_node_modules_folder` function in cli/args/mod.rs (https://github.com/denoland/deno/blob/d1b2a9822d499def513c5b178927c4e05c39caff/cli/args/mod.rs#L1746)

This requires having access to the deno dir at this point though. I think we could:

1. Move the `DenoDirProvider` struct out of cli/factory.rs and instead create it within `CliOptions::new` and store it on `CliOptions`.
2. Expose `DenoDirProvider` on `CliOptions`.
3. Pass `DenoDirProvider` into the `resolve_node_modules_folder` function and then use it in that function.
4. Get and canonicalize the deno dir path inside `resolve_node_modules_folder` in order to compare it against the canonicalized node_modules dir path.

kt3k commented:
> 1. Move the `DenoDirProvider` struct out of cli/factory.rs and instead create it within `CliOptions::new` and store it on `CliOptions`.
> 2. Expose `DenoDirProvider` on `CliOptions`.
> 3. Pass `DenoDirProvider` into the `resolve_node_modules_folder` function and then use it in that function.
> 4. Get and canonicalize the deno dir path inside `resolve_node_modules_folder` in order to compare it against the canonicalized node_modules dir path.

Addressed the above points.

Confirmed `deno run -A npm:create-docusaurus` now works with managed npm resolution.

```shellsession
$ ./target/debug/deno run -A npm:create-docusaurus
✔ What should we name this site? … website
✔ Select a template below... › classic (recommended)
✔ Which language do you want to use? › JavaScript
[INFO] Creating new Docusaurus project...
✔ Select a package manager... › npm
[INFO] Installing dependencies with npm...
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported

added 1212 packages, and audited 1213 packages in 1m

298 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
[SUCCESS] Created website.
```

kt3k commented on cli/args/mod.rs:
```diff
@@ -810,11 +812,14 @@ impl CliOptions {
     let maybe_lockfile = maybe_lockfile.filter(|_| !force_global_cache);
     let root_folder = start_dir.workspace.root_folder_configs();
+    let deno_dir_provider =
+      Arc::new(DenoDirProvider::new(flags.cache_path.clone()));
```
`deno_dir_provider` is not `Deferred` anymore. The `Deferred` defined in `cli/factory.rs` uses `once_cell::unsync::OnceCell` and not thread-safe, that is inconvenient for `CliOptions`. Should I create `once_cell::sync::OnceCell`-version of `Deferred` and wrap this `deno_dir_provider`? (That feels a bit overkill to me for now)

dsherret replied:
Nope. This is good. Thanks!

dsherret commented on cli/args/mod.rs:
```diff
@@ -1776,7 +1784,18 @@ fn resolve_node_modules_folder(
-    // always auto-discover the local_node_modules_folder when a package.json exists
+    if let Ok(deno_dir) = deno_dir_provider.get_or_create() {
+      // `deno_dir.root` can be symlink in macOS
+      if let Ok(root) = canonicalize_path_maybe_not_exists(&deno_dir.root) {
```
I wonder if we should surface these errors when they happen? I guess they'll be surfaced later anyway.

dsherret reviewed (approved):
Great. LGTM!
